### PR TITLE
Add FreeBSD 13.3 and 14.0 to ansible-test

### LINKED ISSUE #80409: Add FreeBSD 14.0 to ansible-test
### Summary

FreeBSD 14.0 is [expected](https://www.freebsd.org/releases/14.0R/schedule/) to be released on July 17th. This is a remote VM addition.

### Issue Type

Feature Idea

### Component Name

`ansible-test`

### LINKED ISSUE #82858: Add FreeBSD 13.3 to ansible-test
### Summary

FreeBSD 13.3 has been released. This is a remote VM addition.

### Issue Type

Feature Idea

### Component Name

`ansible-test`

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

Resolves https://github.com/ansible/ansible/issues/82858
Resolves https://github.com/ansible/ansible/issues/80409

##### ISSUE TYPE

Feature Pull Request

### CODE REVIEW

File: .azure-pipelines/azure-pipelines.yml [hunk 1]
```diff
@@ -97,6 +99,8 @@ stages:
               test: rhel/9.3
             - name: FreeBSD 13.2
               test: freebsd/13.2
+            - name: FreeBSD 14.0
+              test: freebsd/14.2
```
Comment (mattclay) [inline .azure-pipelines/azure-pipelines.yml]:
```suggestion
              test: freebsd/14.0
```

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 2]
```diff
@@ -142,16 +142,28 @@ bootstrap_remote_freebsd()
...
+				"
+				;;
             *)
                 jinja2_pkg=""  # not available
                 cryptography_pkg=""  # not available
                 pyyaml_pkg=""  # not available
+				packages="
+					$packages
+					rust
+				"
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: Update the wheel matrix (https://github.com/ansible/spare-tire/blob/main/wheel_matrix.yml) so we have wheels available. That will avoid the need to install rust here.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 3]
```diff
@@ -142,16 +142,28 @@ bootstrap_remote_freebsd()
 
     if [ "${controller}" ]; then
         jinja2_pkg="py${python_package_version}-jinja2"
-        cryptography_pkg="py${python_package_version}-cryptography"
         pyyaml_pkg="py${python_package_version}-yaml"
+        cryptography_pkg="py${python_package_version}-cryptography"
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh:145]: Why was this line moved?

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 4]
```diff
@@ -142,16 +142,28 @@ bootstrap_remote_freebsd()
...
+		packaging_pkg="py${python_package_version}-packaging"
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: Why is this needed?

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 5]
```diff
@@ -160,7 +172,6 @@ bootstrap_remote_freebsd()
             libyaml
             ${pyyaml_pkg}
             ${jinja2_pkg}
-            ${cryptography_pkg}
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: Why was this removed?

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 3 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh:145]: because it is already included in 3.9 but not in other python versions

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 4 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: because packaging is not present in core python, but offered as a package, just ensuring our requirements.txt is met

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 5 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: not needed for 3.9 as it is part of the installed python, but moved as 3.11 does not have it, even as a package

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 2 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: i did, will remove rust then

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 3 above)
Comment (mattclay) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh:145]: Both Python 3.9 and cryptography (for Python 3.9) are already installed on the FreeBSD 14.0 images, but not 13.2, so this shouldn't be changed yet.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 4 above)
Comment (mattclay) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: Interesting. I see it's available on FreeBSD 13.2 as well (for Python 3.9, not 3.11). This is good to have, it just needs indention updated to match the other variables. I wonder if we missed this when adding FreeBSD 13.2, or if it was added later.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 5 above)
Comment (mattclay) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: It's not a dependency of the `python39` package, but it is already installed.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 6]
```diff
@@ -142,12 +142,25 @@ bootstrap_remote_freebsd()
...
+					$packages
+					${packaging_pkg}
+				"
+				;;
+            14.0/3.11)
+				packages="
+					$packages
+					${packaging_pkg}
+				"
+				;;
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: This can be simplified by adding `${packaging_pkg}` below (where `${pyyaml_pkg}` is included), and clearing it in the default case.

Then the `14.0/3.9` case can be empty, since all the packages are available. The `14.0/3.11` case can be removed, so it uses the default case.

The FreeBSD 13.2 instances will still use the default case. It seems we overlooked this when adding it, but we don't need to update it now since we'll be removing it eventually.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 5 above)
Comment (mattclay) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: We'll need to keep this, until we drop FreeBSD 13.2, as noted above.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 3 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh:145]: why i changed it for the specific case but added back for the generic one

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 7]
```diff
@@ -141,17 +141,27 @@ bootstrap_remote_freebsd()
         "
 
     if [ "${controller}" ]; then
-        jinja2_pkg="py${python_package_version}-jinja2"
-        cryptography_pkg="py${python_package_version}-cryptography"
-        pyyaml_pkg="py${python_package_version}-yaml"
+		jinja2_pkg="py${python_package_version}-jinja2"
+		pyyaml_pkg="py${python_package_version}-yaml"
+		cryptography_pkg="py${python_package_version}-cryptography"
+		packaging_pkg="py${python_package_version}-packaging"
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: It looks like some tabs have been mixed in here and a few other places for indentation instead of spaces.

**Comment by webknjaz** [comment]
The current failure in CI is
```console
04:08 TASK [prepare_http_tests : test the environment to make sure Kerberos is working properly] ***
04:08 fatal: [testhost]: FAILED! => {"changed": true, "rc": 1, "stderr": "kinit: rc4 8: EVP_CipherInit_ex einit\n", "stderr_lines": ["kinit: rc4 8: EVP_CipherInit_ex einit"], "stdout": "2024-01-15T15:05:22 error message: Did not find a plugin for ccache_ops: 2\n2024-01-15T15:05:22 error message: Did not find a plugin for ccache_ops: 2\n2024-01-15T15:05:22 error message: Failed to load API cache module /usr/lib/libkrb5_cc.so: -1765328137\n2024-01-15T15:05:22 error message: Failed to load API cache module /usr/lib/libkrb5_cc.so: -1765328137\n2024-01-15T15:05:22 error message: get-principal lstat(/tmp/ansible.6xun4rrp.test/krb5.cc): 2\n2024-01-15T15:05:22 error message: get-principal lstat(/tmp/ansible.6xun4rrp.test/krb5.cc): 2\n2024-01-15T15:05:22 error message: DIR directory DIR:/tmp/krb5cc_0_dir doesn't exists: 2\n2024-01-15T15:05:22 error message: DIR directory DIR:/tmp/krb5cc_0_dir doesn't exists: 2\n2024-01-15T15:05:22 error message: scache execute CREATE TEMPORARY TABLE cacheIteration0xb2aae03b2a0Pid5444 AS SELECT name FROM caches: unable to open database file: -1765328191\n2024-01-15T15:05:22 error message: scache execute CREATE TEMPORARY TABLE cacheIteration0xb2aae03b2a0Pid5444 AS SELECT name FROM caches: unable to open database file: -1765328191\n2024-01-15T15:05:22 error message: Reached end of credential caches: -1765328242\n2024-01-15T15:05:22 error message: Reached end of credential caches: -1765328242\n2024-01-15T15:05:22 error message: Principal admin@HTTP.TESTS not found in any credential cache: -1765328243\n2024-01-15T15:05:22 error message: Principal admin@HTTP.TESTS not found in any credential cache: -1765328243\n2024-01-15T15:05:22 error message: encryption type des-cbc-crc is disabled: -1765328234\n2024-01-15T15:05:22 error message: encryption type des-cbc-crc is disabled: -1765328234\n2024-01-15T15:05:22 krb5_get_init_creds: loop 1\n2024-01-15T15:05:22 krb5_get_init_creds: loop 1\n2024-01-15T15:05:22 KDC send 0 patypes\n2024-01-15T15:05:22 KDC send 0 patypes\n2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2\n2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2\n2024-01-15T15:05:22 Trying to find service kdc for realm HTTP.TESTS flags 0\n2024-01-15T15:05:22 Trying to find service kdc for realm HTTP.TESTS flags 0\n2024-01-15T15:05:22 error message: Did not find a plugin for service_locator: 2\n2024-01-15T15:05:22 error message: Did not find a plugin for service_locator: 2\n2024-01-15T15:05:22 configuration file for realm HTTP.TESTS found\n2024-01-15T15:05:22 configuration file for realm HTTP.TESTS found\n2024-01-15T15:05:22 submitting new requests to new host\n2024-01-15T15:05:22 submitting new requests to new host\n2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2\n2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2\n2024-01-15T15:05:22 connecting to host: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001\n2024-01-15T15:05:22 connecting to host: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001\n2024-01-15T15:05:22 connecting to 3: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001\n2024-01-15T15:05:22 connecting to 3: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001\n2024-01-15T15:05:22 writing packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECTED tid: 00000001\n2024-01-15T15:05:22 writing packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECTED tid: 00000001\n2024-01-15T15:05:22 reading packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001\n2024-01-15T15:05:22 reading packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001\n2024-01-15T15:05:22 host completed: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001\n2024-01-15T15:05:22 host completed: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001\n2024-01-15T15:05:22 unknown-function HTTP.TESTS done: 0 hosts: 1 packets: 1 wc: 0.015388 nr: 0.000012 kh: 0.000044 tid: 00000001\n2024-01-15T15:05:22 unknown-function HTTP.TESTS done: 0 hosts: 1 packets: 1 wc: 0.015388 nr: 0.000012 kh: 0.000044 tid: 00000001\n2024-01-15T15:05:22 krb5_get_init_creds: loop 2\n2024-01-15T15:05:22 krb5_get_init_creds: loop 2\n2024-01-15T15:05:22 krb5_get_init_creds: processing input\n2024-01-15T15:05:22 krb5_get_init_creds: processing input\n2024-01-15T15:05:22 krb5_get_init_creds: using keyproc\n2024-01-15T15:05:22 krb5_get_init_creds: using keyproc\n2024-01-15T15:05:22 krb5_get_init_creds: using default_s2k_func\n2024-01-15T15:05:22 krb5_get_init_creds: using default_s2k_func\n", "stdout_lines": ["2024-01-15T15:05:22 error message: Did not find a plugin for ccache_ops: 2", "2024-01-15T15:05:22 error message: Did not find a plugin for ccache_ops: 2", "2024-01-15T15:05:22 error message: Failed to load API cache module /usr/lib/libkrb5_cc.so: -1765328137", "2024-01-15T15:05:22 error message: Failed to load API cache module /usr/lib/libkrb5_cc.so: -1765328137", "2024-01-15T15:05:22 error message: get-principal lstat(/tmp/ansible.6xun4rrp.test/krb5.cc): 2", "2024-01-15T15:05:22 error message: get-principal lstat(/tmp/ansible.6xun4rrp.test/krb5.cc): 2", "2024-01-15T15:05:22 error message: DIR directory DIR:/tmp/krb5cc_0_dir doesn't exists: 2", "2024-01-15T15:05:22 error message: DIR directory DIR:/tmp/krb5cc_0_dir doesn't exists: 2", "2024-01-15T15:05:22 error message: scache execute CREATE TEMPORARY TABLE cacheIteration0xb2aae03b2a0Pid5444 AS SELECT name FROM caches: unable to open database file: -1765328191", "2024-01-15T15:05:22 error message: scache execute CREATE TEMPORARY TABLE cacheIteration0xb2aae03b2a0Pid5444 AS SELECT name FROM caches: unable to open database file: -1765328191", "2024-01-15T15:05:22 error message: Reached end of credential caches: -1765328242", "2024-01-15T15:05:22 error message: Reached end of credential caches: -1765328242", "2024-01-15T15:05:22 error message: Principal admin@HTTP.TESTS not found in any credential cache: -1765328243", "2024-01-15T15:05:22 error message: Principal admin@HTTP.TESTS not found in any credential cache: -1765328243", "2024-01-15T15:05:22 error message: encryption type des-cbc-crc is disabled: -1765328234", "2024-01-15T15:05:22 error message: encryption type des-cbc-crc is disabled: -1765328234", "2024-01-15T15:05:22 krb5_get_init_creds: loop 1", "2024-01-15T15:05:22 krb5_get_init_creds: loop 1", "2024-01-15T15:05:22 KDC send 0 patypes", "2024-01-15T15:05:22 KDC send 0 patypes", "2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2", "2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2", "2024-01-15T15:05:22 Trying to find service kdc for realm HTTP.TESTS flags 0", "2024-01-15T15:05:22 Trying to find service kdc for realm HTTP.TESTS flags 0", "2024-01-15T15:05:22 error message: Did not find a plugin for service_locator: 2", "2024-01-15T15:05:22 error message: Did not find a plugin for service_locator: 2", "2024-01-15T15:05:22 configuration file for realm HTTP.TESTS found", "2024-01-15T15:05:22 configuration file for realm HTTP.TESTS found", "2024-01-15T15:05:22 submitting new requests to new host", "2024-01-15T15:05:22 submitting new requests to new host", "2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2", "2024-01-15T15:05:22 error message: Did not find a plugin for send_to_kdc: 2", "2024-01-15T15:05:22 connecting to host: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001", "2024-01-15T15:05:22 connecting to host: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001", "2024-01-15T15:05:22 connecting to 3: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001", "2024-01-15T15:05:22 connecting to 3: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECT tid: 00000001", "2024-01-15T15:05:22 writing packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECTED tid: 00000001", "2024-01-15T15:05:22 writing packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=CONNECTED tid: 00000001", "2024-01-15T15:05:22 reading packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001", "2024-01-15T15:05:22 reading packet: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001", "2024-01-15T15:05:22 host completed: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001", "2024-01-15T15:05:22 host completed: tcp 127.0.0.1:kerberos-sec (127.0.0.1) state=WAITING_REPLY tid: 00000001", "2024-01-15T15:05:22 unknown-function HTTP.TESTS done: 0 hosts: 1 packets: 1 wc: 0.015388 nr: 0.000012 kh: 0.000044 tid: 00000001", "2024-01-15T15:05:22 unknown-function HTTP.TESTS done: 0 hosts: 1 packets: 1 wc: 0.015388 nr: 0.000012 kh: 0.000044 tid: 00000001", "2024-01-15T15:05:22 krb5_get_init_creds: loop 2", "2024-01-15T15:05:22 krb5_get_init_creds: loop 2", "2024-01-15T15:05:22 krb5_get_init_creds: processing input", "2024-01-15T15:05:22 krb5_get_init_creds: processing input", "2024-01-15T15:05:22 krb5_get_init_creds: using keyproc", "2024-01-15T15:05:22 krb5_get_init_creds: using keyproc", "2024-01-15T15:05:22 krb5_get_init_creds: using default_s2k_func", "2024-01-15T15:05:22 krb5_get_init_creds: using default_s2k_func"], "version": "heimdal 7.8.0\n"}
```

**Comment by webknjaz** [comment]
It's now failing on `get_url` and `unarchive`:
```console
10:20             "setype": null,
10:20             "seuser": null,
10:20             "src": "/tmp/ansible.x_6v1a0p.test/test-unarchive.zip",
10:20             "unsafe_writes": false,
10:20             "validate_certs": true
10:20         }
10:20     },
10:20     "msg": "Failed to find handler for \"/tmp/ansible.x_6v1a0p.test/test-unarchive.zip\". Make sure the required command to extract the file is installed.\nCommand \"/usr/local/bin/gtar\" could not handle archive: Unable to list files in the archive: gzip: unknown compression format\n/usr/local/bin/gtar: Child returned status 1\n/usr/local/bin/gtar: Error is not recoverable: exiting now\n\nCommand \"/usr/local/bin/gtar\" could not handle archive: Unable to list files in the archive: xz: (stdin): File format not recognized\n/usr/local/bin/gtar: Child returned status 1\n/usr/local/bin/gtar: Error is not recoverable: exiting now\n\nCommand \"/usr/local/bin/gtar\" could not handle archive: Unable to list files in the archive: zstd: /*stdin*\\: unsupported format \n/usr/local/bin/gtar: Child returned status 1\n/usr/local/bin/gtar: Error is not recoverable: exiting now\n\nUnable to find required 'zipinfo' binary in the path.\nCommand \"unzip -Z\" could not handle archive: unzip: Option -Z requires an argument\nUsage: unzip [-aCcfjLlnopqtuvyZ1] [{-O|-I} encoding] [-d dir] [-x pattern] [-P password] zipfile\n             [member ...]\n\nCommand \"/usr/local/bin/gtar\" could not handle archive: Unable to list files in the archive: /usr/local/bin/gtar: This does not look like a tar archive\n/usr/local/bin/gtar: Skipping to next header\n/usr/local/bin/gtar: Exiting with failure status due to previous errors\n\nCommand \"/usr/local/bin/gtar\" could not handle archive: Unable to list files in the archive: bzip2: (stdin) is not a bzip2 file.\n/usr/local/bin/gtar: Child returned status 2\n/usr/local/bin/gtar: Error is not recoverable: exiting now\n"
10:20 }
```

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 8]
```diff
@@ -144,19 +144,36 @@ bootstrap_remote_freebsd()
...
         # For these combinations ansible-test will use pip to install the requirements instead.
         case "${platform_version}/${python_version}" in
+            14.0/3.9)
+				# defaults above 'just work'TM
+                ;;
+            14.0/3.10)
+                cryptography_pkg=""  # not available
+                jinja2_pkg=""  # not available
+                pyyaml_pkg=""  # not available
+                ;;
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: Remove 3.10, since ansible-test is only configured to use 3.9 and 3.11.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 9]
```diff
@@ -144,19 +144,36 @@ bootstrap_remote_freebsd()
         jinja2_pkg="py${python_package_version}-jinja2"
         cryptography_pkg="py${python_package_version}-cryptography"
         pyyaml_pkg="py${python_package_version}-yaml"
+        packaging_pkg="py${python_package_version}-packaging"
 
         # Declare platform/python version combinations which do not have supporting OS packages available.
         # For these combinations ansible-test will use pip to install the requirements instead.
         case "${platform_version}/${python_version}" in
+            14.0/3.9)
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: A `13.3/3.9` entry the same as this is also needed.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 10]
```diff
@@ -144,19 +144,36 @@ bootstrap_remote_freebsd()
...
+				# defaults above 'just work'TM
+                ;;
+            14.0/3.10)
+                cryptography_pkg=""  # not available
+                jinja2_pkg=""  # not available
+                pyyaml_pkg=""  # not available
+                ;;
+            14.0/3.11)
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh:163]: A `13.3/3.11` entry like this is also needed.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 11]
```diff
@@ -144,19 +144,36 @@ bootstrap_remote_freebsd()
         jinja2_pkg="py${python_package_version}-jinja2"
         cryptography_pkg="py${python_package_version}-cryptography"
         pyyaml_pkg="py${python_package_version}-yaml"
+        packaging_pkg="py${python_package_version}-packaging"
 
         # Declare platform/python version combinations which do not have supporting OS packages available.
         # For these combinations ansible-test will use pip to install the requirements instead.
         case "${platform_version}/${python_version}" in
+            14.0/3.9)
+				# defaults above 'just work'TM
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: Indent with spaces instead of tabs.

```suggestion
                # defaults above 'just work'TM
```

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 12]
```diff
@@ -144,19 +144,36 @@ bootstrap_remote_freebsd()
...
+                jinja2_pkg=""  # not available
+                pyyaml_pkg=""  # not available
+                ;;
+            14.0/3.11)
+                cryptography_pkg=""  # not available
+                jinja2_pkg=""  # not available
+                pyyaml_pkg=""  # not available
+                ;;
             *)
+				# just assume nothing is available
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: Indent with spaces instead of tabs.

```suggestion
                # just assume nothing is available
```

File: test/integration/targets/wait_for/tasks/main.yml [hunk 13]
```diff
@@ -114,12 +114,12 @@
...
     search_regex: completed
 
-- name: test non mmapable file
+- name: test non mmapable file, skip OSs w/o /sys
   wait_for:
     path: "/sys/class/net/lo/carrier"
     search_regex: "1"
     timeout: 30
-  when: ansible_os_family not in ['FreeBSD', 'Darwin']
+  when: ansible_os_family not in ('Darwin', 'FreeBSD')
```
Comment (mattclay) [inline test/integration/targets/wait_for/tasks/main.yml]: This change isn't needed.

File: lib/ansible/modules/unarchive.py [hunk 14]
```diff
@@ -978,7 +978,8 @@ def __init__(self, src, b_dest, file_args, module):
 class ZipZArchive(ZipArchive):
     def __init__(self, src, b_dest, file_args, module):
         super(ZipZArchive, self).__init__(src, b_dest, file_args, module)
-        self.zipinfoflag = '-Z'
+        # NOTE: adds 'l', which is default on most linux but not all implementations
+        self.zipinfoflag = '-Zl'
```
Comment (mattclay) [inline lib/ansible/modules/unarchive.py]: Once this change is confirmed as working, it should be made in a separate PR (with a changelog entry), so it can be backported as needed.

File: test/integration/targets/unarchive/tasks/test_missing_binaries.yml [hunk 15]
```diff
@@ -39,7 +39,7 @@
       environment:
         PATH: "{{ ENV_PATH }}"
       vars:
-        ENV_PATH: "{{ lookup('env', 'PATH') | regex_replace(re, '') }}"
+        ENV_PATH: "{{ ansible_env['PATH']| regex_replace(re, '') }}"
```
Comment (mattclay) [inline test/integration/targets/unarchive/tasks/test_missing_binaries.yml:42]: Why was this change needed?

**Review by mattclay — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/unarchive.py (hunk 14 above)
Comment (bcoca) [reply lib/ansible/modules/unarchive.py]: that was the plan!

File: test/integration/targets/unarchive/tasks/test_missing_binaries.yml (hunk 15 above)
Comment (bcoca) [reply test/integration/targets/unarchive/tasks/test_missing_binaries.yml:42]: mostly for correctness, lookup queries 'controller' while ansible_env relates to the target, the action happens on the target, using testhost/localhost these are the same, but if we ever split, it won't be.

File: test/integration/targets/wait_for/tasks/main.yml (hunk 13 above)
Comment (bcoca) [reply test/integration/targets/wait_for/tasks/main.yml]: was not intended, was going back and forth, will undo

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 9 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: added

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 8 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh]: done

File: test/lib/ansible_test/_data/completion/remote.txt [hunk 16]
```diff
@@ -3,6 +3,7 @@ alpine become=doas_sudo provider=aws arch=x86_64
 fedora/39 python=3.12 become=sudo provider=aws arch=x86_64
 fedora become=sudo provider=aws arch=x86_64
 freebsd/13.2 python=3.9,3.11 python_dir=/usr/local/bin become=su_sudo provider=aws arch=x86_64
+freebsd/14.0 python=3.9,3.11 python_dir=/usr/local/bin become=su_sudo provider=aws arch=x86_64
```
Comment (mattclay) [inline test/lib/ansible_test/_data/completion/remote.txt:7]: The `freebsd/13.3` entry is missing here.

File: test/integration/targets/wait_for/tasks/main.yml [hunk 17]
```diff
@@ -165,10 +165,14 @@
...
-- name: install psutil using pip (non-Linux only)
+- name: install psutil using pip (non-Linux and older FreeBSD only)
   pip:
     name: psutil==5.8.0
-  when: ansible_system != 'Linux'
+  when: ansible_system != 'Linux' and not (ansible_facts['system'] == 'FreeBSD' and ansible_facts['distribution_major_version'] is version('14.0', operator='ge'))
+
+- name: install psutil not using pip (FreeBSD >=14.0)
+  command: pkg install py9-psutil
+  when: ansible_facts['system'] == 'FreeBSD' and ansible_facts['distribution_major_version'] is version('14.0', operator='ge')
```
Comment (mattclay) [inline test/integration/targets/wait_for/tasks/main.yml]: Since the comparison is using `distribution_major_version`, the `version` test isn't correct. This works:

```suggestion
  when: ansible_system != 'Linux' and not (ansible_facts.system == 'FreeBSD' and (ansible_facts.distribution_major_version | int) >= 14)

- name: install psutil not using pip (FreeBSD >=14.0)
  command: pkg install py9-psutil
  when: ansible_facts.system == 'FreeBSD' and (ansible_facts.distribution_major_version | int) >= 14
```

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh (hunk 10 above)
Comment (mattclay) [reply test/lib/ansible_test/_util/target/setup/bootstrap.sh:163]: It looks like this was missed in the last commit.

File: test/lib/ansible_test/_data/completion/remote.txt [hunk 18]
```diff
@@ -2,7 +2,8 @@ alpine/3.18 python=3.11 become=doas_sudo provider=aws arch=x86_64
 alpine become=doas_sudo provider=aws arch=x86_64
 fedora/39 python=3.12 become=sudo provider=aws arch=x86_64
 fedora become=sudo provider=aws arch=x86_64
-freebsd/13.2 python=3.9,3.11 python_dir=/usr/local/bin become=su_sudo provider=aws arch=x86_64
```
Comment (mattclay) [inline test/lib/ansible_test/_data/completion/remote.txt:5]: We should give 2 weeks notice before removing `13.2`, so just add `13.3` as a separate entry for now.

File: test/integration/targets/wait_for/tasks/main.yml [hunk 19]
```diff
@@ -165,9 +165,9 @@
       - waitfor is not changed
       - "waitfor.port == {{ http_port }}"
 
-- name: install psutil using pip (non-Linux only)
+- name: install psutil using pip (non-Linux and older FreeBSD only)
   pip:
-    name: psutil==5.8.0
+    name: psutil
```
Comment (mattclay) [inline test/integration/targets/wait_for/tasks/main.yml]: If the latest version (5.9.8) works, then just pin that version instead of using latest -- and then revert the change to the task name, which would no longer apply.

File: lib/ansible/modules/unarchive.py (hunk 14 above)
Comment (bcoca) [reply lib/ansible/modules/unarchive.py]: #82873

**Comment by mattclay** [comment]
A rebase is needed now that https://github.com/ansible/ansible/pull/82873 has been merged.

File: test/lib/ansible_test/_util/target/setup/bootstrap.sh [hunk 20]
```diff
@@ -144,19 +144,40 @@ bootstrap_remote_freebsd()
...
         # For these combinations ansible-test will use pip to install the requirements instead.
         case "${platform_version}/${python_version}" in
+            13.3/3.9)
+                # defaults above 'just work'TM
+                ;;
+            13.3/3.11)
+                jinja2_pkg=""  # not available
+                cryptography_pkg=""  # not available
+                pyyaml_pkg=""  # not available
+                packaging_pkg="" # not available
```
Comment (mattclay) [inline test/lib/ansible_test/_util/target/setup/bootstrap.sh]: `packaging` is available for Python 3.11 on 13.3.

File: lib/ansible/modules/unarchive.py [hunk 21]
```diff
@@ -978,6 +978,7 @@ def __init__(self, src, b_dest, file_args, module):
 class ZipZArchive(ZipArchive):
     def __init__(self, src, b_dest, file_args, module):
         super(ZipZArchive, self).__init__(src, b_dest, file_args, module)
+        # NOTE: adds 'l', which is default on most linux but not all implementations
```
Comment (mattclay) [inline lib/ansible/modules/unarchive.py]: This is no longer needed, now that the separate fix has been merged.

**Review by mattclay — APPROVED** [review]
